infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/utils/UpgradeManager.java

```diff
@@ -0,0 +1,213 @@
+    //  reflect the version per node.
+    public static final String VERSION_PLUGIN_KEY = "VERSION";
+
+    private final VersionString versionString = VersionString.newBuilder().setName(VERSION_PLUGIN_KEY).build();
+
+    private final CorfuStore corfuStore;
+
+    @Getter
+    private LRRollingUpgradeHandler lrRollingUpgradeHandler;
+
+    private ILogReplicationVersionAdapter logReplicationVersionAdapter;
+
+    private final String pluginConfigFilePath;
+
+    @Getter
+    private static String currentVersion;
+
+    // TODO: Metadata type need not be uuid.  It can be a generic protobuf message.
+    private Table<VersionString, Version, Uuid> pluginVersionTable;
+
+    public UpgradeManager(CorfuRuntime runtime, String pluginConfigFilePath) {
+        this.pluginConfigFilePath = pluginConfigFilePath;
+        this.corfuStore = new CorfuStore(runtime);
+        initVersionPlugin(runtime);
+        setupVersionTable();
+        initLogReplicationRollingUpgradeHandler(corfuStore);
+    }
+
+    /**
+     * Instantiate the LogReplicator's Rolling Upgrade Handler and invoke its
+     * check the first time, so it can cache the result in the common case
+     * where there is no rolling upgrade in progress.
+     *
+     * @param corfuStore - instance of the store to which the check is made with.
+     */
+    private void initLogReplicationRollingUpgradeHandler(CorfuStore corfuStore) {
+        this.lrRollingUpgradeHandler = new LRRollingUpgradeHandler(logReplicationVersionAdapter);
+        final int retries = 3;
+        for (int i = retries; i>=0; i--) {
+            try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+                log.info("LRRollingUpgradeHandler: Prestart check isUpgradeOn: {}",
+                    lrRollingUpgradeHandler.isLRUpgradeInProgress(txn));
+                txn.commit();
+                break;
+            } catch (Exception ex) {
+                log.error("Fatal error: Failed to get LR upgrade status", ex);
+            }
+        }
+    }
+
+    private void initVersionPlugin(CorfuRuntime runtime) {
+        log.info("Version plugin :: {}", pluginConfigFilePath);
+        LogReplicationPluginConfig config = new LogReplicationPluginConfig(pluginConfigFilePath);
+        File jar = new File(config.getStreamFetcherPluginJARPath());
+        try (URLClassLoader child = new URLClassLoader(new URL[]{jar.toURI().toURL()}, this.getClass().getClassLoader())) {
+            Class plugin = Class.forName(config.getStreamFetcherClassCanonicalName(), true, child);
+            logReplicationVersionAdapter = (ILogReplicationVersionAdapter)
+                plugin.getDeclaredConstructor(CorfuRuntime.class).newInstance(runtime);
+        } catch (Exception e) {
+            log.error("Fatal error: Failed to get Log Replicator Version Plugin", e);
+            throw new UnrecoverableCorfuError(e);
+        }
+    }
```

Comment:
This has been there with ConfigManager for while, but it just occurs to me that the line 
```
LogReplicationPluginConfig config = new LogReplicationPluginConfig(pluginConfigFilePath);
```
will always seek to `DefaultLogReplicationConfigAdapter` which gives  a version string "LATEST-VERSION" upon any error case. But the fact that we throws an UnrecoverableCorfuError means that we want to terminate LR in this case right?

So do we want to terminate LR if the given `pluginConfigFilePath` is invalid (doesn't provide a valid version string)? If so I think there might be some changes needed in `LogReplicationPluginConfig`, otherwise even the verticals will get a "LATEST-VERSION" as their version string in error cases.